test/distributed/org/apache/cassandra/distributed/upgrade/MixedModeBatchTestBase.java

```diff
@@ -0,0 +1,155 @@
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.distributed.upgrade;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.cassandra.distributed.UpgradeableCluster;
+import org.apache.cassandra.distributed.api.ConsistencyLevel;
+import org.apache.cassandra.distributed.api.IMessageFilters;
+import org.apache.cassandra.distributed.shared.Versions;
+import org.apache.cassandra.exceptions.WriteFailureException;
+import org.apache.cassandra.exceptions.WriteTimeoutException;
+
+import static org.apache.cassandra.distributed.shared.AssertUtils.assertRows;
+import static org.apache.cassandra.distributed.shared.AssertUtils.row;
+import static org.apache.cassandra.net.Verb.BATCH_STORE_REQ;
+import static org.apache.cassandra.net.Verb.REQUEST_RSP;
+
+public class MixedModeBatchTestBase extends UpgradeTestBase
+{
+    private static final int KEYS_PER_BATCH = 10;
+
+    protected void testSimpleStrategy(Versions.Major from, Versions.Major to, boolean isLogged) throws Throwable
+    {
+        String insert = "INSERT INTO test_simple.names (key, name) VALUES (%d, '%s')";
+        String select = "SELECT * FROM test_simple.names WHERE key = ?";
+
+        new TestCase()
+        .nodes(3)
+        .nodesToUpgrade(1, 2)
+        .upgrade(from, to)
+        .setup(cluster -> {
+            cluster.schemaChange("CREATE KEYSPACE test_simple WITH replication = {'class': 'SimpleStrategy', 'replication_factor': 2};");
+            cluster.schemaChange("CREATE TABLE test_simple.names (key int PRIMARY KEY, name text)");
+        })
+        .runAfterNodeUpgrade((cluster, upgraded) -> {
+            if (isLogged)
+            {
+                // If we're testing logged batches, exercise the case were batchlog writes fail.
+                IMessageFilters.Filter dropBatchlogWrite = cluster.filters().inbound().verbs(BATCH_STORE_REQ.id, REQUEST_RSP.id).drop();
+                dropBatchlogWrite.on();
+                testBatches(isLogged, true, insert, select, cluster, upgraded);
+                cluster.filters().reset();
+            }
+
+            cluster.coordinator(1).execute("TRUNCATE test_simple.names", ConsistencyLevel.ALL);
+            testBatches(isLogged, false, insert, select, cluster, upgraded);
+            cluster.coordinator(1).execute("TRUNCATE test_simple.names", ConsistencyLevel.ALL);
+        })
+        .run();
+    }
+
+    private void testBatches(boolean isLogged, boolean failBatchlog, String insert, String select, UpgradeableCluster cluster, int upgraded)
+    {
+        List<Long> initialTokens = new ArrayList<>(cluster.size() + 1);
+        initialTokens.add(null); // The first valid token is at 1 to avoid offset math below.
+
+        for (int i = 1; i <= cluster.size(); i++)
+            initialTokens.add(Long.valueOf(cluster.get(i).config().get("initial_token").toString()));
+
+        List<Long> validTokens = initialTokens.subList(1, cluster.size() + 1);
```

Comment:
I think we don't need to put the `null` there, especially when we are removing it in the later call to `subList` to prepare a list valid for `primaryReplica`. We could just add them directly to the same list that we are going to pass to `primaryReplica`:
```suggestion
        List<Long> initialTokens = new ArrayList<>(cluster.size());
        for (int i = 1; i <= cluster.size(); i++)
            initialTokens.add(Long.valueOf(cluster.get(i).config().get("initial_token").toString()));
```
If anything, we could specify in the JavaDoc for `primaryReplica` that the parameter uses zero-based numbering but the result is one-based.